File: src/beaker_notebook/subkernels/rlang/procedures/reflectors/data_frame.r
{# target_type: data.frame #}
{# function_name: ._beaker_reflect_data_frame #}
._beaker_reflect_data_frame <- function(.name, .value) {
    .rows <- nrow(.value)
    .cols <- ncol(.value)
    .shape <- paste0(.rows, "x", .cols)
    .col_classes <- sapply(.value, function(c) class(c)[1])
    .sample <- tryCatch({
        .out <- capture.output(print(utils::head(.value, 3)))
        .joined <- paste(.out, collapse = "\n")
        if (nchar(.joined) > 400) {
            list(substr(.joined, 1, 400), TRUE)
        } else {
            list(.joined, .rows > 3)
        }
    }, error = function(e) list("<unrepresentable>", FALSE))
    list(
        tag = paste0("data.frame[", .shape, "]"),
        description = list(
            type = "data.frame",
            shape = .shape,
            size = .rows,
            summary = list(columns = as.list(.col_classes)),
            sample = .sample[[1]],
            truncated = .sample[[2]]
        )
    )
}

File: src/beaker_notebook/subkernels/python/procedures/reflectors/dataframe.py
{# target_type: pandas.core.frame.DataFrame #}
{# function_name: _beaker_reflect_dataframe #}
def _beaker_reflect_dataframe(_name, _value):
    _rows, _cols = _value.shape
    _shape = f"{_rows}x{_cols}"
    _tag = f"DataFrame[{_shape}]"

    _dtypes = {str(_col): str(_dt) for _col, _dt in _value.dtypes.items()}
    try:
        _sample = _value.head(3).to_string()
    except Exception:
        _sample = "<unrepresentable>"
    if len(_sample) > 400:
        _sample = _sample[:400]
        _truncated = True
    else:
        _truncated = _rows > 3
    return {
        "tag": _tag,
        "description": {
            "type": "pandas.DataFrame",
            "shape": _shape,
            "size": int(_rows),
            "summary": {"columns": _dtypes},
            "sample": _sample,
            "truncated": _truncated,
        },
    }
